From f143397edcad1a76e33876cfb5bfa0b51ae17d50 Mon Sep 17 00:00:00 2001
From: Vighnesh Maheshwari <vighmah@amazon.com>
Date: Thu, 16 Oct 2025 10:00:47 -0700
Subject: [PATCH] openssl-util: build without ui.h

Remove some code that depends on openssl/ui.h which is not provided by
aws-lc. This can probably be submitted upstream.

[Refreshed for v257.13: Extended patch to guard new UI_METHOD references
in openssl-util.c (load_key_from_provider, load_key_from_engine, and
openssl_load_private_key) that were added upstream since 257.9.]

Signed-off-by: Vighnesh Maheshwari <vighmah@amazon.com>
Signed-off-by: Maher Homsi <maherhom@amazon.com>
---
 src/shared/openssl-util.c | 17 ++++++++++++++---
 src/shared/openssl-util.h | 11 +++++++++++
 2 files changed, 25 insertions(+), 3 deletions(-)

diff --git a/src/shared/openssl-util.c b/src/shared/openssl-util.c
index a48bb1c..8d3e9ff 100644
--- a/src/shared/openssl-util.c
+++ b/src/shared/openssl-util.c
@@ -1319,7 +1319,11 @@ int pkey_generate_volume_keys(
 static int load_key_from_provider(
                 const char *provider,
                 const char *private_key_uri,
+#ifndef OPENSSL_NO_UI_CONSOLE
                 UI_METHOD *ui_method,
+#else
+                void *ui_method,
+#endif /* OPENSSL_NO_UI_CONSOLE */
                 EVP_PKEY **ret) {
 
         assert(provider);
@@ -1362,7 +1366,13 @@ static int load_key_from_provider(
 #endif
 }
 
-static int load_key_from_engine(const char *engine, const char *private_key_uri, UI_METHOD *ui_method, EVP_PKEY **ret) {
+static int load_key_from_engine(const char *engine, const char *private_key_uri,
+#ifndef OPENSSL_NO_UI_CONSOLE
+                UI_METHOD *ui_method,
+#else
+                void *ui_method,
+#endif /* OPENSSL_NO_UI_CONSOLE */
+                EVP_PKEY **ret) {
         assert(engine);
         assert(private_key_uri);
         assert(ret);
@@ -1659,9 +1669,10 @@ int openssl_load_private_key(
                 if (r < 0)
                         return log_debug_errno(r, "Failed to allocate ask-password user interface: %m");
 
-                UI_METHOD *ui_method = NULL;
 #ifndef OPENSSL_NO_UI_CONSOLE
-                ui_method = ui->method;
+                UI_METHOD *ui_method = ui->method;
+#else
+                void *ui_method = NULL;
 #endif
 
                 switch (private_key_source_type) {
diff --git a/src/shared/openssl-util.h b/src/shared/openssl-util.h
index e25a175..784e764 100644
--- a/src/shared/openssl-util.h
+++ b/src/shared/openssl-util.h
@@ -41,6 +41,9 @@ int parse_openssl_key_source_argument(const char *argument, char **private_key_s
 #  ifndef OPENSSL_NO_UI_CONSOLE
 #    include <openssl/ui.h>
 #  endif
+#  ifndef OPENSSL_HMAC_H
+#    include <openssl/hmac.h>
+#  endif
 #  include <openssl/x509v3.h>
 #  ifndef OPENSSL_VERSION_MAJOR
 /* OPENSSL_VERSION_MAJOR macro was added in OpenSSL 3. Thus, if it doesn't exist,  we must be before OpenSSL 3. */
@@ -151,7 +154,9 @@ int digest_and_sign(const EVP_MD *md, EVP_PKEY *privkey, const void *data, size_
 typedef struct X509 X509;
 typedef struct EVP_PKEY EVP_PKEY;
 typedef struct EVP_MD EVP_MD;
+#  ifndef OPENSSL_NO_UI_CONSOLE
 typedef struct UI_METHOD UI_METHOD;
+#  endif
 typedef struct ASN1_TYPE ASN1_TYPE;
 typedef struct ASN1_STRING ASN1_STRING;

@@ -182,10 +187,16 @@ DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(EVP_PKEY*, EVP_PKEY_free, NULL);
 DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(ASN1_TYPE*, ASN1_TYPE_free, NULL);
 DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(ASN1_STRING*, ASN1_STRING_free, NULL);

+#ifndef OPENSSL_NO_UI_CONSOLE
 struct OpenSSLAskPasswordUI {
         AskPasswordRequest request;
         UI_METHOD *method;
 };
+#else
+struct OpenSSLAskPasswordUI {
+        AskPasswordRequest request;
+};
+#endif /* OPENSSL_NO_UI_CONSOLE */

 OpenSSLAskPasswordUI* openssl_ask_password_ui_free(OpenSSLAskPasswordUI *ui);
